README.md
[![WPOnion Framework Screenshot](http://s3.wponion.com/wp.org/v3/banner-1544x500.jpg)](http://s3.wponion.com/wp.org/v3/banner-1544x500.jpg)

[![All Contributors](https://img.shields.io/badge/all_contributors-2-orange.svg?style=flat-square)](#contributors)

## So what is WPOnion , exactly?

WPOnion is a simple yet powerful framework that helps developers build custom meta boxes and custom fields in WordPress fast and easily.

The framework lets you define custom meta boxes and custom fields via arrays and handles everything behind the scene automatically. It has a wide range of field types, field settings and supports not only post meta but also term meta, user meta, comment meta, settings pages and custom tables.

## Installation

### Stable 
The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

[![Latest Stable Version](https://poser.pugx.org/wponion/wponion/version)](https://packagist.org/packages/wponion/wponion)
[![Total Downloads](https://poser.pugx.org/wponion/wponion/downloads)](https://packagist.org/packages/wponion/wponion)
[![WP](https://img.shields.io/badge/WordPress-Standar-1abc9c.svg)](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/)
[![License](https://poser.pugx.org/wponion/wponion/license)](https://packagist.org/packages/wponion/wponion)

To install **WPOnion library**, simply:

    $ composer require wponion/wponion

The previous command will only install the necessary files, if you prefer to **download the entire source code** you can use:

    $ composer require wponion/wponion --prefer-source

You can also **clone the complete repository** with Git:

    $ git clone https://github.com/wponion/wponion.git

---

### Development 
The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

To install **WPOnion library**, simply:

    $ composer require wponion/wponion:dev-development

The previous command will only install the necessary files, if you prefer to **download the entire source code** you can use:

    $ composer require wponion/wponion:dev-development --prefer-source

You can also **clone the complete repository** with Git:

    $ git clone https://github.com/wponion/wponion.git

---

## Usage 
***If Downloaded Manually***
Add the below code your plugin / theme file 

```php
require __DIR__.'/wponion/wponion.php';

add_action('wponion_loaded','your-prefix_after_wponion_loaded');

function your-prefix_after_wponion_loaded(){
    //do your stuff here.
}
```

## Contribution

Like WPOnion Framework, this documentation is open sourced on Github & GitBooks. The website's content is written in Markdown.

If you find anything incorrect in the documentation or out-dated, please help us fix it.

In case you come across anything along the way that we haven’t covered, or if you know of a tip you think others would find handy, please file an issue and we’ll see about including it in this guide.

## Documentation

https://wponion.gitbook.io/docs/

### Screenshots / Demo

https://demo.wponion.com

## Contributors
Thanks to all the Contributors who worked on this project to get it done asap
<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
<table><tr><td align="center"><a href="https://EnterpriseBranding.io"><img src="https://avatars3.githubusercontent.com/u/6123260?v=4" width="100px;" alt="Enterprise Branding"/><br /><sub><b>Enterprise Branding</b></sub></a><br /><a href="https://github.com/wponion/wponion/commits?author=EnterpriseBranding" title="Documentation">📖</a> <a href="#ideas-EnterpriseBranding" title="Ideas, Planning, & Feedback">🤔</a></td><td align="center"><a href="https://redpandaventures.ltd"><img src="https://avatars1.githubusercontent.com/u/909468?v=4" width="100px;" alt="digitalchild"/><br /><sub><b>digitalchild</b></sub></a><br /><a href="#ideas-digitalchild" title="Ideas, Planning, & Feedback">🤔</a></td></tr></table>

<!-- ALL-CONTRIBUTORS-LIST:END -->

## License
This project is licensed under **MIT**. See the [LICENSE](LICENSE) file for more info.

## Backed By
| [![DigitalOcean][do-image]][do-ref] | [![JetBrains][jb-image]][jb-ref] |  [![Tidio Chat][tidio-image]][tidio-ref] |
| --- | --- | --- |

[do-image]: https://vsp.ams3.cdn.digitaloceanspaces.com/cdn/DO_Logo_Horizontal_Blue-small.png
[jb-image]: https://vsp.ams3.cdn.digitaloceanspaces.com/cdn/phpstorm-small.png?v3
[tidio-image]: https://vsp.ams3.cdn.digitaloceanspaces.com/cdn/tidiochat-small.png
[do-ref]: https://s.svarun.in/Ef
[jb-ref]: https://www.jetbrains.com
[tidio-ref]: https://tidiochat.com


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),

and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


# [1.5.3.7] - 25/03/2021
## Fixed
* Admin Notice Module Not working
* Improved Tippyjs integration and avoided most Tippyjs warnings
* Some jQuery Migrate Warning fixed 
* Spinner field css fixed.

## Changed
* Bump [sweetalert2] from `10.15.5` to `10.15.6` #359
* Bump [@popperjs/core] from `2.9.0` to `2.9.1` #355
* Bump [@wordpress/hooks] from `2.11.1` to `2.12.0` #358
* [Security] Bump [yargs-parser] from `5.0.0` to `5.0.1` #356

# [1.5.3.6] - 08/03/2021
## Added
* Option to bulk add fields in nested fields

## Fixed
* Color Picker Field (many issues with its possible options)
* WC Product Metabox Ajax Loading Issue

# [1.5.3.5] - 28/02/2021
## Fixed
* Fixed Dashboard Widget Loading if no fields are set
* Fixed FAQ Field Content Area
* Fixed PHP Warning when saving settings for the first time.

## Changed
* Bump [sweetalert2] from `10.14.0` to `10.15.5` #342
* Bump [jquery-serializejson] from `3.2.0` to `3.2.1` #350
* Bump [tippy.js] from `6.2.7` to `6.3.1` #351
* Bump [@popperjs/core] from `2.6.0` to `2.9.0` #354

# [1.5.3.4] - 27/02/2021
## Fixed
* Tab Field CSS Issue
* Settings Page Ajax Saving Issue

# [1.5.3.3] - 08/02/2021
## Changed
* Bump [wordpress-js-ports] from `1.0.9` to `1.0.11` #287 
* Bump [tippy.js] from `6.2.5` to `6.2.7` #291 
* Bump [@popperjs/core] from `2.4.4` to `2.6.0` #319 
* Bump [jquery-serializejson] from `2.9.0` to `3.2.0` #323 
* Bump [bootstrap] from `4.5.0` to `4.6.0` #327
* Bump [sweetalert2] from `9.17.0` to `10.14.0` #331 
* Bump [@wordpress/hooks] from `2.9.0` to `2.11.1` #332 
* __[Security]__ Bump ini from `1.3.5` to `1.3.8` #318 
* __[Security]__ Bump jquery-validation from `1.19.2` to `1.19.3` #326
* __[Security]__ Bump node-sass from `4.12.0` to `4.14.1` #270
* __[Security]__ Bump elliptic from `6.5.0` to `6.5.4` #333 
* Support Files (Icons, Google Fonts) Updated

---

# [1.5.3.2] - 23/07/2020
## Added
* DB Save Hook Support For Settings & Network Settings Module
* `set_unique` function in **Unique** Trait
* `set_module` function in **Module** Trait
* new `object` Trait created
* Custom Save Handler For Every Module.
* **Action :** `wponion/metabox/render/before`
* **Action :** `wponion/metabox/render/{$this->unique()}/before`
* **Action :** `wponion/metabox/render/{$this->unique()}/after`
* **Action :** `wponion/metabox/render/after`
* Custom Class Option In Admin Page Module
* Notification Bubble Support In Admin Page Module
* Menu Separator Support In Admin Page Module
 
## Fixed
* Taxonmy Register Issue . fixed by changing `$this->taxonomy` to `$this->unique()`
* **CSS** Tab layout `z-index`
* Ajax issue with metabox saving

## Changed
* Bump [@wordpress/hooks] from `2.8.0` to `2.9.0`
* Bump [@popperjs/core] from `2.4.2` to `2.4.4`
* Bump [tippy.js] from `6.2.3` to `6.2.5`
* Bump [sweetalert2] from `9.15.1` to `9.17.0`
* Bump [css-checkbox-library] from `1.5.1` to `1.5.2`

## Removed
* Deleted local function calls added by mistake


# [1.5.3.1] - 23/06/2020
## Fixed
* Custom Template Tags Not Working In Selectbox

# [1.5.3] - 17/06/2020
## Added
* Option To Customize Page's Browser Title

## Fixed
* Select Ajax Query Args Not Working Fields Outside WPOnion Module 

## Changed
* Bump [sweetalert2] from `9.14.4` to `9.15.1`
* WPOnion Core Assets CDN Version To `1.4`

# [1.5.2] - 14/06/2020
## Fixed 
* [Fatal Error When Using Along Side Elementor Plugin](https://github.com/wponion/wponion/issues/252)

## Changed
* Bump [sweetalert2] from `9.14.0` to `9.14.4` #250 
* Bump [@popperjs/core] from `2.4.0` to `2.4.2` #251 

# [1.5.1] - 06/06/2020

## Changed
* `wponion_icon` function now can handle image link

## Fixed
* Performace Fix in Storage Interface.

# [1.5] - 06/06/2020
## Added
* **Hook Priority** option in taxonomy
* `WPONION_PLUGIN_FILE` constant in plugin file.
* Deprecation handlers for (Function,Actions,Filters)
* Nested Array Support for `Class_Options` Trait
* Option to have external links in **Admin Page** Module
* Changes Browser URL when click tab in settings page (useful if everything loads in a single page)
* is empty check with array `wponion_is_array`
* Created Storage Handlers which interacts with wordpress DB.
* **Action**   :  "wponion/{$this->module()}/get/before"
* **Action**   :  "wponion/{$this->module()}/{$this->unique()}/get/before"
* **Action**   :  "wponion/{$this->module()}/{$this->unique()}/get/after"
* **Action**   :  "wponion/{$this->module()}/get/after"
* **Action**   :  "wponion/{$this->module()}/save/before"
* **Action**   :  "wponion/{$this->module()}/{$this->unique()}/save/before"
* **Action**   :  "wponion/{$this->module()}/{$this->unique()}/save/after"
* **Action**   :  "wponion/{$this->module()}/save/after"
* **Filters**  :  "wponion/{$this->module()}/get/values"
* **Filters**  :  "wponion/{$this->module()}/{$this->unique()}/get/values"
* **Filters**  :  "wponion/{$this->module()}/save/values"
* **Filters**  :  "wponion/{$this->module()}/{$this->unique()}//save/values"
* **Function** :  `wponion_cast_array`
* **Function** :  `wponion_has_column_class`
* **Function** :  `wponion_parse_args_mixed`
* **Function** :  `wponion_parse_args_forced_values`
* **Function** :  `wponion_cast_array`
* **Function** :  `wponion_define`
* **Function** :  `wponion_is_defined`
* **Function** :  `wponion_is_set`
* **Function** :  `wponion_is_version`

## Fixed
* [#233  `theme` arg in Taxonomy Module produces error if theme is not `wp_modern`](https://github.com/wponion/wponion/issues/233)
* CSS fix for switcher field.
* Admin Notice Database Fetching Issue.
* Assets Loading issue in `Page Actions` Module
* Improved Performance.

## Changed
* Bump [sweetalert2] from `9.10.12` to `9.14.0`
* Reworked Admin Notice Module and changed all function names to match wponion format

fields/card.php
<?php
/**
 *
 * Initial version created 12-05-2018 / 07:17 PM
 *
 * @author Varun Sridharan <varunsridharan23@gmail.com>
 * @version 1.0
 * @since 1.0
 * @package
 * @link
 * @copyright 2018 Varun Sridharan
 * @license GPLV3 Or Greater (https://www.gnu.org/licenses/gpl-3.0.txt)
 */

namespace WPOnion\Field;
if ( ! defined( 'ABSPATH' ) ) {
	die;
}

if ( ! class_exists( '\WPOnion\Field\card' ) ) {
	/**
	 * Class card
	 *
	 * @package WPOnion\Field
	 * @author Varun Sridharan <varunsridharan23@gmail.com>
	 * @since 1.0
	 */
	class card extends \WPOnion\Field {
		/**
		 * Final HTML Output;
		 *
		 * @return mixed;
		 */
		protected function output() {
			if ( ! $this->has( 'options' ) ) {
				return false;
			}

			echo '<div class="row">';
			foreach ( $this->data( 'options' ) as $option ) {
				$option             = $this->parse_args( $option, array(
					'image'           => false,
					'content'         => false,
					'body'            => false,
					'wrap_attributes' => false,
					'wrap_class'      => false,
				) );
				$wrap_attr          = $option['wrap_attributes'];
				$wrap_attr['class'] = wponion_html_class( $wrap_attr['class'], array( $option['wrap_class'], 'card' ) );

				echo '<div class="' . $this->data( 'card_col' ) . '">';
				echo '<div ' . wponion_array_to_html_attributes( $wrap_attr ) . '>';
				if ( false !== $option['image'] ) {
					$image = $this->handle_data( $option['image'], array(
						'src'        => false,
						'attributes' => array(),
						'alt'        => false,
					), 'src' );

					if ( ! isset( $image['attributes']['src'] ) ) {
						$image['attributes']['src'] = $image['src'];
					}

					if ( ! isset( $image['attributes']['alt'] ) ) {
						$image['attributes']['alt'] = $image['alt'];
					}

					echo '<img ' . wponion_array_to_html_attributes( $image['attributes'] ) . '/>';
				}

				if ( false !== $option['body'] || false !== $option['content'] ) {
					echo '<div class="card-body">';
					echo $option['body'];
					echo $option['content'];
					echo '</div>';
				}

				echo '</div>';
				echo '</div>';
			}

			echo '</div>';
		}

		/**
		 * Loads the required plugins assets.
		 *
		 * @return mixed|void
		 */
		public function field_assets() {
			// TODO: Implement field_assets() method.
		}

		/**
		 * Returns all fields default.
		 *
		 * @return array|mixed
		 */
		protected function field_default() {
			return array(
				'options'  => array(),
				'card_col' => 'col',

			);
		}
	}
}
